### ISSUE TITLE
ZFS 0.6.3-1 on CentOS 6.6 hangs some time after booting

### ISSUE THREAD

**Issue by Pentium100MHz** [zfsonlinux/zfs#3000]
Hello,

One of my servers hangs randomly some time after booting, the system drives work OK, but zfs dataset hangs (that is, any attempt to access it results in a stuck process). I updated to the latest version of CentOS and zfs, but the issue remained.

The server makes use of snapshots. The last time this problem happened was after destroying a bunch of snapshots (could be a coincidence).

zfs version:        0.6.3-1.2
spl version:        0.6.3-1.2
kernel version: 2.6.32-504.3.3.el6.x86_64

There are no error messages in syslog, other than the ones informing that a task (that is trying to access the zfs dataset) is blocked for more than 120 seconds.

Running iotop I see that [txg_sync] is using ~30% IO, but has zero read/write B/s and 
"mkdir -p /pool/dataset/dir1" (I tried to create a folder) is using ~500B/s and 0.5% IO. 

Iostat shows all zpool drives as completely idle (zeros in all columns). dd if=/dev/sdx of=/dev/null works fine.

**Comment by behlendorf** [comment]
Can you please include the back trace from the console.

**Comment by Pentium100MHz** [comment]
If/when the server locks up again, I will do so.

**Comment by Pentium100MHz** [comment]
The server hanged again:
Here is the trace:
```
INFO: task rsync:39792 blocked for more than 120 seconds.
      Tainted: P           ---------------    2.6.32-504.3.3.el6.x86_64 #1
"echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
rsync         D 000000000000000c     0 39792  39791 0x00000080
 ffff88016df9fb58 0000000000000086 0000000000000000 ffff880000036c20
 ffff88017790eaa0 00000000000412d0 00020312f446e027 ffffffff81133d53
 ffffffff8100bb8e 0000000121c0a1de ffff88017790f058 ffff88016df9ffd8
Call Trace:
 [<ffffffff81133d53>] ? __alloc_pages_nodemask+0x113/0x8d0
 [<ffffffff8100bb8e>] ? apic_timer_interrupt+0xe/0x20
 [<ffffffff8105859f>] ? mutex_spin_on_owner+0x9f/0xc0
 [<ffffffff8152b306>] __mutex_lock_slowpath+0x96/0x210
 [<ffffffff811a9537>] ? __d_lookup+0xa7/0x150
 [<ffffffff8152ae2b>] mutex_lock+0x2b/0x50
 [<ffffffff8119dbdb>] do_lookup+0x11b/0x230
 [<ffffffff8119e8f4>] __link_path_walk+0x7a4/0x1000
 [<ffffffff8119f40a>] path_walk+0x6a/0xe0
 [<ffffffff8119f61b>] filename_lookup+0x6b/0xc0
 [<ffffffff811a0747>] user_path_at+0x57/0xa0
 [<ffffffff8104d18c>] ? __do_page_fault+0x1ec/0x480
 [<ffffffff81193984>] ? cp_new_stat+0xe4/0x100
 [<ffffffff81193bc0>] vfs_fstatat+0x50/0xa0
 [<ffffffff81193c7e>] vfs_lstat+0x1e/0x20
 [<ffffffff81193ca4>] sys_newlstat+0x24/0x50
 [<ffffffff810e5c87>] ? audit_syscall_entry+0x1d7/0x200
 [<ffffffff8100b072>] system_call_fastpath+0x16/0x1b
INFO: task rsync:39792 blocked for more than 120 seconds.
      Tainted: P           ---------------    2.6.32-504.3.3.el6.x86_64 #1
"echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
rsync         D 000000000000000c     0 39792  39791 0x00000080
 ffff88016df9fb58 0000000000000086 0000000000000000 ffff880000036c20
 ffff88017790eaa0 00000000000412d0 00020312f446e027 ffffffff81133d53
 ffffffff8100bb8e 0000000121c0a1de ffff88017790f058 ffff88016df9ffd8
Call Trace:
 [<ffffffff81133d53>] ? __alloc_pages_nodemask+0x113/0x8d0
 [<ffffffff8100bb8e>] ? apic_timer_interrupt+0xe/0x20
 [<ffffffff8105859f>] ? mutex_spin_on_owner+0x9f/0xc0
 [<ffffffff8152b306>] __mutex_lock_slowpath+0x96/0x210
 [<ffffffff811a9537>] ? __d_lookup+0xa7/0x150
 [<ffffffff8152ae2b>] mutex_lock+0x2b/0x50
 [<ffffffff8119dbdb>] do_lookup+0x11b/0x230
 [<ffffffff8119e8f4>] __link_path_walk+0x7a4/0x1000
 [<ffffffff8119f40a>] path_walk+0x6a/0xe0
 [<ffffffff8119f61b>] filename_lookup+0x6b/0xc0
 [<ffffffff811a0747>] user_path_at+0x57/0xa0
 [<ffffffff8104d18c>] ? __do_page_fault+0x1ec/0x480
 [<ffffffff81193984>] ? cp_new_stat+0xe4/0x100
 [<ffffffff81193bc0>] vfs_fstatat+0x50/0xa0
 [<ffffffff81193c7e>] vfs_lstat+0x1e/0x20
 [<ffffffff81193ca4>] sys_newlstat+0x24/0x50
 [<ffffffff810e5c87>] ? audit_syscall_entry+0x1d7/0x200
 [<ffffffff8100b072>] system_call_fastpath+0x16/0x1b
INFO: task rsync:39792 blocked for more than 120 seconds.
      Tainted: P           ---------------    2.6.32-504.3.3.el6.x86_64 #1
"echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
rsync         D 000000000000000c     0 39792  39791 0x00000080
 ffff88016df9fb58 0000000000000086 0000000000000000 ffff880000036c20
 ffff88017790eaa0 00000000000412d0 00020312f446e027 ffffffff81133d53
 ffffffff8100bb8e 0000000121c0a1de ffff88017790f058 ffff88016df9ffd8
Call Trace:
 [<ffffffff81133d53>] ? __alloc_pages_nodemask+0x113/0x8d0
 [<ffffffff8100bb8e>] ? apic_timer_interrupt+0xe/0x20
 [<ffffffff8105859f>] ? mutex_spin_on_owner+0x9f/0xc0
 [<ffffffff8152b306>] __mutex_lock_slowpath+0x96/0x210
 [<ffffffff811a9537>] ? __d_lookup+0xa7/0x150
 [<ffffffff8152ae2b>] mutex_lock+0x2b/0x50
 [<ffffffff8119dbdb>] do_lookup+0x11b/0x230
 [<ffffffff8119e8f4>] __link_path_walk+0x7a4/0x1000
 [<ffffffff8119f40a>] path_walk+0x6a/0xe0
 [<ffffffff8119f61b>] filename_lookup+0x6b/0xc0
 [<ffffffff811a0747>] user_path_at+0x57/0xa0
 [<ffffffff8104d18c>] ? __do_page_fault+0x1ec/0x480
 [<ffffffff81193984>] ? cp_new_stat+0xe4/0x100
 [<ffffffff81193bc0>] vfs_fstatat+0x50/0xa0
 [<ffffffff81193c7e>] vfs_lstat+0x1e/0x20
 [<ffffffff81193ca4>] sys_newlstat+0x24/0x50
 [<ffffffff810e5c87>] ? audit_syscall_entry+0x1d7/0x200
 [<ffffffff8100b072>] system_call_fastpath+0x16/0x1b
INFO: task rsync:39925 blocked for more than 120 seconds.
      Tainted: P           ---------------    2.6.32-504.3.3.el6.x86_64 #1
"echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
rsync         D 0000000000000002     0 39925  39924 0x00000080
 ffff880100e7db58 0000000000000086 ffff880100e7c000 0000000000012240
 00000000000284d0 0000000000000002 ffff88013dc67170 000280da00000000
 ffffffff8100bb8e ffff880100e7db58 ffff880122a15af8 ffff880100e7dfd8
Call Trace:
 [<ffffffff8100bb8e>] ? apic_timer_interrupt+0xe/0x20
 [<ffffffff8105858d>] ? mutex_spin_on_owner+0x8d/0xc0
 [<ffffffff8152b306>] __mutex_lock_slowpath+0x96/0x210
 [<ffffffff811a9537>] ? __d_lookup+0xa7/0x150
 [<ffffffff8152ae2b>] mutex_lock+0x2b/0x50
 [<ffffffff8119dbdb>] do_lookup+0x11b/0x230
 [<ffffffff8119e8f4>] __link_path_walk+0x7a4/0x1000
 [<ffffffff8114f0b7>] ? handle_pte_fault+0xf7/0xb00
 [<ffffffff8119f40a>] path_walk+0x6a/0xe0
 [<ffffffff8119f61b>] filename_lookup+0x6b/0xc0
 [<ffffffff811a0747>] user_path_at+0x57/0xa0
 [<ffffffff8104d18c>] ? __do_page_fault+0x1ec/0x480
 [<ffffffff81193984>] ? cp_new_stat+0xe4/0x100
 [<ffffffff81193bc0>] vfs_fstatat+0x50/0xa0
 [<ffffffff81193c7e>] vfs_lstat+0x1e/0x20
 [<ffffffff81193ca4>] sys_newlstat+0x24/0x50
 [<ffffffff810e5c87>] ? audit_syscall_entry+0x1d7/0x200
 [<ffffffff8100b072>] system_call_fastpath+0x16/0x1b
INFO: task rsync:39792 blocked for more than 120 seconds.
      Tainted: P           ---------------    2.6.32-504.3.3.el6.x86_64 #1
"echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
rsync         D 000000000000000c     0 39792  39791 0x00000080
 ffff88016df9fb58 0000000000000086 0000000000000000 ffff880000036c20
 ffff88017790eaa0 00000000000412d0 00020312f446e027 ffffffff81133d53
 ffffffff8100bb8e 0000000121c0a1de ffff88017790f058 ffff88016df9ffd8
Call Trace:
 [<ffffffff81133d53>] ? __alloc_pages_nodemask+0x113/0x8d0
 [<ffffffff8100bb8e>] ? apic_timer_interrupt+0xe/0x20
 [<ffffffff8105859f>] ? mutex_spin_on_owner+0x9f/0xc0
 [<ffffffff8152b306>] __mutex_lock_slowpath+0x96/0x210
 [<ffffffff811a9537>] ? __d_lookup+0xa7/0x150
 [<ffffffff8152ae2b>] mutex_lock+0x2b/0x50
 [<ffffffff8119dbdb>] do_lookup+0x11b/0x230
 [<ffffffff8119e8f4>] __link_path_walk+0x7a4/0x1000
 [<ffffffff8119f40a>] path_walk+0x6a/0xe0
 [<ffffffff8119f61b>] filename_lookup+0x6b/0xc0
 [<ffffffff811a0747>] user_path_at+0x57/0xa0
 [<ffffffff8104d18c>] ? __do_page_fault+0x1ec/0x480
 [<ffffffff81193984>] ? cp_new_stat+0xe4/0x100
 [<ffffffff81193bc0>] vfs_fstatat+0x50/0xa0
 [<ffffffff81193c7e>] vfs_lstat+0x1e/0x20
 [<ffffffff81193ca4>] sys_newlstat+0x24/0x50
 [<ffffffff810e5c87>] ? audit_syscall_entry+0x1d7/0x200
 [<ffffffff8100b072>] system_call_fastpath+0x16/0x1b
INFO: task rsync:39925 blocked for more than 120 seconds.
      Tainted: P           ---------------    2.6.32-504.3.3.el6.x86_64 #1
"echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
rsync         D 0000000000000002     0 39925  39924 0x00000080
 ffff880100e7db58 0000000000000086 ffff880100e7c000 0000000000012240
 00000000000284d0 0000000000000002 ffff88013dc67170 000280da00000000
 ffffffff8100bb8e ffff880100e7db58 ffff880122a15af8 ffff880100e7dfd8
Call Trace:
 [<ffffffff8100bb8e>] ? apic_timer_interrupt+0xe/0x20
 [<ffffffff8105858d>] ? mutex_spin_on_owner+0x8d/0xc0
 [<ffffffff8152b306>] __mutex_lock_slowpath+0x96/0x210
 [<ffffffff811a9537>] ? __d_lookup+0xa7/0x150
 [<ffffffff8152ae2b>] mutex_lock+0x2b/0x50
 [<ffffffff8119dbdb>] do_lookup+0x11b/0x230
 [<ffffffff8119e8f4>] __link_path_walk+0x7a4/0x1000
 [<ffffffff8114f0b7>] ? handle_pte_fault+0xf7/0xb00
 [<ffffffff8119f40a>] path_walk+0x6a/0xe0
 [<ffffffff8119f61b>] filename_lookup+0x6b/0xc0
 [<ffffffff811a0747>] user_path_at+0x57/0xa0
 [<ffffffff8104d18c>] ? __do_page_fault+0x1ec/0x480
 [<ffffffff81193984>] ? cp_new_stat+0xe4/0x100
 [<ffffffff81193bc0>] vfs_fstatat+0x50/0xa0
 [<ffffffff81193c7e>] vfs_lstat+0x1e/0x20
 [<ffffffff81193ca4>] sys_newlstat+0x24/0x50
 [<ffffffff810e5c87>] ? audit_syscall_entry+0x1d7/0x200
 [<ffffffff8100b072>] system_call_fastpath+0x16/0x1b
INFO: task rsync:39792 blocked for more than 120 seconds.
      Tainted: P           ---------------    2.6.32-504.3.3.el6.x86_64 #1
"echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
rsync         D 000000000000000c     0 39792  39791 0x00000080
 ffff88016df9fb58 0000000000000086 0000000000000000 ffff880000036c20
 ffff88017790eaa0 00000000000412d0 00020312f446e027 ffffffff81133d53
 ffffffff8100bb8e 0000000121c0a1de ffff88017790f058 ffff88016df9ffd8
Call Trace:
 [<ffffffff81133d53>] ? __alloc_pages_nodemask+0x113/0x8d0
 [<ffffffff8100bb8e>] ? apic_timer_interrupt+0xe/0x20
 [<ffffffff8105859f>] ? mutex_spin_on_owner+0x9f/0xc0
 [<ffffffff8152b306>] __mutex_lock_slowpath+0x96/0x210
 [<ffffffff811a9537>] ? __d_lookup+0xa7/0x150
 [<ffffffff8152ae2b>] mutex_lock+0x2b/0x50
 [<ffffffff8119dbdb>] do_lookup+0x11b/0x230
 [<ffffffff8119e8f4>] __link_path_walk+0x7a4/0x1000
 [<ffffffff8119f40a>] path_walk+0x6a/0xe0
 [<ffffffff8119f61b>] filename_lookup+0x6b/0xc0
 [<ffffffff811a0747>] user_path_at+0x57/0xa0
 [<ffffffff8104d18c>] ? __do_page_fault+0x1ec/0x480
 [<ffffffff81193984>] ? cp_new_stat+0xe4/0x100
 [<ffffffff81193bc0>] vfs_fstatat+0x50/0xa0
 [<ffffffff81193c7e>] vfs_lstat+0x1e/0x20
 [<ffffffff81193ca4>] sys_newlstat+0x24/0x50
 [<ffffffff810e5c87>] ? audit_syscall_entry+0x1d7/0x200
 [<ffffffff8100b072>] system_call_fastpath+0x16/0x1b
INFO: task rsync:39925 blocked for more than 120 seconds.
      Tainted: P           ---------------    2.6.32-504.3.3.el6.x86_64 #1
"echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
rsync         D 0000000000000002     0 39925  39924 0x00000080
 ffff880100e7db58 0000000000000086 ffff880100e7c000 0000000000012240
 00000000000284d0 0000000000000002 ffff88013dc67170 000280da00000000
 ffffffff8100bb8e ffff880100e7db58 ffff880122a15af8 ffff880100e7dfd8
Call Trace:
 [<ffffffff8100bb8e>] ? apic_timer_interrupt+0xe/0x20
 [<ffffffff8105858d>] ? mutex_spin_on_owner+0x8d/0xc0
 [<ffffffff8152b306>] __mutex_lock_slowpath+0x96/0x210
 [<ffffffff811a9537>] ? __d_lookup+0xa7/0x150
 [<ffffffff8152ae2b>] mutex_lock+0x2b/0x50
 [<ffffffff8119dbdb>] do_lookup+0x11b/0x230
 [<ffffffff8119e8f4>] __link_path_walk+0x7a4/0x1000
 [<ffffffff8114f0b7>] ? handle_pte_fault+0xf7/0xb00
 [<ffffffff8119f40a>] path_walk+0x6a/0xe0
 [<ffffffff8119f61b>] filename_lookup+0x6b/0xc0
 [<ffffffff811a0747>] user_path_at+0x57/0xa0
 [<ffffffff8104d18c>] ? __do_page_fault+0x1ec/0x480
 [<ffffffff81193984>] ? cp_new_stat+0xe4/0x100
 [<ffffffff81193bc0>] vfs_fstatat+0x50/0xa0
 [<ffffffff81193c7e>] vfs_lstat+0x1e/0x20
 [<ffffffff81193ca4>] sys_newlstat+0x24/0x50
 [<ffffffff810e5c87>] ? audit_syscall_entry+0x1d7/0x200
 [<ffffffff8100b072>] system_call_fastpath+0x16/0x1b
INFO: task rsync:39792 blocked for more than 120 seconds.
      Tainted: P           ---------------    2.6.32-504.3.3.el6.x86_64 #1
"echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
rsync         D 000000000000000c     0 39792  39791 0x00000080
 ffff88016df9fb58 0000000000000086 0000000000000000 ffff880000036c20
 ffff88017790eaa0 00000000000412d0 00020312f446e027 ffffffff81133d53
 ffffffff8100bb8e 0000000121c0a1de ffff88017790f058 ffff88016df9ffd8
Call Trace:
 [<ffffffff81133d53>] ? __alloc_pages_nodemask+0x113/0x8d0
 [<ffffffff8100bb8e>] ? apic_timer_interrupt+0xe/0x20
 [<ffffffff8105859f>] ? mutex_spin_on_owner+0x9f/0xc0
 [<ffffffff8152b306>] __mutex_lock_slowpath+0x96/0x210
 [<ffffffff811a9537>] ? __d_lookup+0xa7/0x150
 [<ffffffff8152ae2b>] mutex_lock+0x2b/0x50
 [<ffffffff8119dbdb>] do_lookup+0x11b/0x230
 [<ffffffff8119e8f4>] __link_path_walk+0x7a4/0x1000
 [<ffffffff8119f40a>] path_walk+0x6a/0xe0
 [<ffffffff8119f61b>] filename_lookup+0x6b/0xc0
 [<ffffffff811a0747>] user_path_at+0x57/0xa0
 [<ffffffff8104d18c>] ? __do_page_fault+0x1ec/0x480
 [<ffffffff81193984>] ? cp_new_stat+0xe4/0x100
 [<ffffffff81193bc0>] vfs_fstatat+0x50/0xa0
 [<ffffffff81193c7e>] vfs_lstat+0x1e/0x20
 [<ffffffff81193ca4>] sys_newlstat+0x24/0x50
 [<ffffffff810e5c87>] ? audit_syscall_entry+0x1d7/0x200
 [<ffffffff8100b072>] system_call_fastpath+0x16/0x1b
INFO: task rsync:39925 blocked for more than 120 seconds.
      Tainted: P           ---------------    2.6.32-504.3.3.el6.x86_64 #1
"echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
rsync         D 0000000000000002     0 39925  39924 0x00000080
 ffff880100e7db58 0000000000000086 ffff880100e7c000 0000000000012240
 00000000000284d0 0000000000000002 ffff88013dc67170 000280da00000000
 ffffffff8100bb8e ffff880100e7db58 ffff880122a15af8 ffff880100e7dfd8
Call Trace:
 [<ffffffff8100bb8e>] ? apic_timer_interrupt+0xe/0x20
 [<ffffffff8105858d>] ? mutex_spin_on_owner+0x8d/0xc0
 [<ffffffff8152b306>] __mutex_lock_slowpath+0x96/0x210
 [<ffffffff811a9537>] ? __d_lookup+0xa7/0x150
 [<ffffffff8152ae2b>] mutex_lock+0x2b/0x50
 [<ffffffff8119dbdb>] do_lookup+0x11b/0x230
 [<ffffffff8119e8f4>] __link_path_walk+0x7a4/0x1000
 [<ffffffff8114f0b7>] ? handle_pte_fault+0xf7/0xb00
 [<ffffffff8119f40a>] path_walk+0x6a/0xe0
 [<ffffffff8119f61b>] filename_lookup+0x6b/0xc0
 [<ffffffff811a0747>] user_path_at+0x57/0xa0
 [<ffffffff8104d18c>] ? __do_page_fault+0x1ec/0x480
 [<ffffffff81193984>] ? cp_new_stat+0xe4/0x100
 [<ffffffff81193bc0>] vfs_fstatat+0x50/0xa0
 [<ffffffff81193c7e>] vfs_lstat+0x1e/0x20
 [<ffffffff81193ca4>] sys_newlstat+0x24/0x50
 [<ffffffff810e5c87>] ? audit_syscall_entry+0x1d7/0x200
 [<ffffffff8100b072>] system_call_fastpath+0x16/0x1b
```

iostat -x 2
```
avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           5.73    0.00    0.08    0.00    0.00   94.19

Device:         rrqm/s   wrqm/s     r/s     w/s   rsec/s   wsec/s avgrq-sz avgqu-sz   await  svctm  %util
sda               0.00     0.00    0.00    1.50     0.00    12.00     8.00     0.00    0.00   0.00   0.00
sdb               0.00     0.00    0.00    0.00     0.00     0.00     0.00     0.00    0.00   0.00   0.00
sdd               0.00     0.00    0.00    0.00     0.00     0.00     0.00     0.00    0.00   0.00   0.00
sde               0.00     0.00    0.00    0.00     0.00     0.00     0.00     0.00    0.00   0.00   0.00
sdc               0.00     0.00    0.00    0.00     0.00     0.00     0.00     0.00    0.00   0.00   0.00
```

sda is system drive, the rest are for zfs.

**Comment by bougui** [comment]
Hi all,

I'm seeing the similar behavior on my Centos 6.6-minimal Box with 22 drives in my zpool + 2 SSD drive for cache.  I found out that if I don't enable the compression on my zpool I end up with very slow behavior and sometime not even able to write to the zpool via NFS. 

But after enabling compression like this
> zfs set compression=lz4 storage

All seem fine, no more blocking for us.

About my setup, ZFS server 32 GB ram Linux on 2 drives and the zpool has 24 SAS drive + 2 SSD.  This server used to be running OpenIndianna without any issue for the last ~ 3 years. 

We use this server to dump Virtual machines from ESX to Centos ZFS via NFS.

Our current kernel is: 2.6.32-504.8.1.el6.x86_64
And zfs is: zfs-0.6.3-1.2.el6

I'm wandering if you see the same behavior on your system....  when I ask for history while dumping data from my ESX to the ZFS server
> zpool history

It takes more than 2 minutes to responds with 5 lines in it, the zpool creation and the 4 dataset in the pool.

zpool status is pretty fast (instantaneous) 

I just saw that the sharenfs option is not enable for this dataset, could that flag explain some of our problem (Since I'm only using the dataset from nfs ! )?

> zfs get all storage | grep nfs

>> storage  sharenfs              off                    default


Thanks for any advise

Guillaume.
